## Костыра Екатерина ПМ_1901

## Занятие 5. Линейные модели для классификации

Веберите любые данные для задачи классификации из репозитория данных для машинного обучения (UCI Machine learning repository: http://archive.ics.uci.edu/ml/index.php) или возьмите свои данные и постройте линейные модели в соответствии со следующей схемой. Комментарии к каждому разделу обязательны.

### Информация об атрибутах:

1. длина чашелистика в см
2. ширина чашелистика в см
3. длина лепестка в см
4. ширина лепестка в см
5. класс:
- Iris Setosa
- Iris Versicolour
- Iris Virginica

Посмотрим на данные

In [7]:
filename = "bezdekIris (2).data"
names = ['Длина чашелистика','Ширина чашелистика','Длина лепестка','Ширина лепестка','Класс']
dataframe = read_csv(filename, names=names)
array = dataframe.values
dataframe.head()

Длина чашелистика  Ширина чашелистика  Длина лепестка  Ширина лепестка  \
0                5.1                 3.5             1.4              0.2   
1                4.9                 3.0             1.4              0.2   
2                4.7                 3.2             1.3              0.2   
3                4.6                 3.1             1.5              0.2   
4                5.0                 3.6             1.4              0.2   

         Класс  
0  Iris-setosa  
1  Iris-setosa  
2  Iris-setosa  
3  Iris-setosa  
4  Iris-setosa

Опишем данные

In [8]:
dataframe.describe()

Длина чашелистика  Ширина чашелистика  Длина лепестка  Ширина лепестка
count         150.000000          150.000000      150.000000       150.000000
mean            5.843333            3.057333        3.758000         1.199333
std             0.828066            0.435866        1.765298         0.762238
min             4.300000            2.000000        1.000000         0.100000
25%             5.100000            2.800000        1.600000         0.300000
50%             5.800000            3.000000        4.350000         1.300000
75%             6.400000            3.300000        5.100000         1.800000
max             7.900000            4.400000        6.900000         2.500000

### Logistic Regression

In [2]:
import os
os.chdir("C:/Users/79811/anaconda3/")

In [5]:
# Logistic Regression Classification
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
filename = "bezdekIris (2).data"
names = ['Длина чашелистика','Ширина чашелистика','Длина лепестка','Ширина лепестка','Класс']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:4]
Y = array[:,4]
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = LogisticRegression(solver='liblinear')
results = cross_val_score(model, X, Y, cv=kfold)
print(results.mean())

0.9533333333333334


По результатам видим, что точность (доля правильных прогнозов) составляет 95%

### Linear Discriminant Analysis

Линейный дискриминантный анализ.

Алгоритм включает разработку вероятностной модели для каждого класса на основе конкретного распределения наблюдений для каждой входной переменной. Затем новый пример классифицируется путем вычисления условной вероятности его принадлежности к каждому классу и выбора класса с наибольшей вероятностью.

Как таковая, это относительно простая модель вероятностной классификации, которая делает сильные предположения о распределении каждой входной переменной.


In [4]:
# LDA Classification
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
filename = "bezdekIris (2).data"
names = ['Длина чашелистика','Ширина чашелистика','Длина лепестка','Ширина лепестка','Класс']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:4]
Y = array[:,4]
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = LinearDiscriminantAnalysis()
results = cross_val_score(model, X, Y, cv=kfold)
print(results.mean())

0.9800000000000001


По результатам видим, что точность модели составляется 98%

### Support Vector Machines

Машины опорных векторов (SVM) - это набор контролируемых методов обучения, используемых для классификации , регрессии и обнаружения выбросов. SVM предлагает очень высокую точность по сравнению с другими классификаторами, такими как логистическая регрессия и деревья решений. 

Основная идея SVM - найти максимальную маргинальную гиперплоскость (MMH), которая наилучшим образом разделяет набор данных на классы. Гиперплоскость - это плоскость принятия решений, которая разделяет набор объектов, имеющих различную принадлежность к классам

In [11]:
# SVM Classification
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.svm import SVC
filename = "bezdekIris (2).data"
names = ['Длина чашелистика','Ширина чашелистика','Длина лепестка','Ширина лепестка','Класс']
dataframe = read_csv(filename, names=names)
array = dataframe.values
X = array[:,0:4]
Y = array[:,4]
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = SVC()
results = cross_val_score(model, X, Y, cv=kfold)
print(results.mean())

0.9600000000000002


Точность равна 96%, что говорит о том, что этот метод и вправду работает лучше, чем Logistic Regression